IndexOutOfBoundsException in implicit search

```scala
object Test {
  case class Tuple2K[H[_], T[_], X](h: H[X], t: T[X])

  trait TC[A]

  implicit def case1[F[_]](implicit t: => TC[F[Any]]): TC[Tuple2K[[_] => Any, F, Any]] = ???
  implicit def case2[A, F[_]](implicit r: TC[F[Any]]): TC[A] = ???

  implicitly[TC[Int]] // java.lang.IndexOutOfBoundsException: 0
}
```

Trying to minimize this I hit an other exception which should be easier to understand:
```scala
object Test {
  trait TC[A]

  implicit def case1[F[_]](implicit t: => TC[F[Any]]): TC[String] = ???
  implicit def case2[G[_]](implicit r: TC[G[Any]]): TC[Int] = ???

  implicitly[TC[Int]] // java.lang.IndexOutOfBoundsException: 0
}
```
```scala
Exception in thread "main" java.lang.AssertionError: assertion failed: self instantiation of F, constraint = Constraint(
 uninstVars = F;
 constrained types = 
  [G <: [_$2] => Any](implicit r: => Test.TC[G[Any]]): Test.TC[Int]
, [F <: [_$1] => Any](implicit t: => Test.TC[F[Any]]): Test.TC[String]
 bounds = 
     G := F
     F >: F <: F
 ordering = 
)
```
<details>
<summary>Stack trace</summary>
<pre>
        at scala.Predef$.assert(Predef.scala:219)
        at dotty.tools.dotc.core.Types$TypeVar.instantiateWith(Types.scala:3336)
        at dotty.tools.dotc.core.Types$TypeVar.instantiate(Types.scala:3352)
        at dotty.tools.dotc.typer.Inferencing$.$anonfun$interpolateUndetVars$9(Inferencing.scala:264)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:52)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
        at dotty.tools.dotc.typer.Inferencing$.$anonfun$interpolateUndetVars$2(Inferencing.scala:261)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Inferencing$.interpolate$1(Inferencing.scala:221)
        at dotty.tools.dotc.typer.Inferencing$.interpolateUndetVars(Inferencing.scala:267)
        at dotty.tools.dotc.typer.Typer.$anonfun$adapt$3(Typer.scala:1864)
        at dotty.tools.dotc.reporting.trace$.apply(trace.scala:23)
        at dotty.tools.dotc.typer.Typer.$anonfun$adapt$1(Typer.scala:1862)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:1862)
        at dotty.tools.dotc.typer.Typer.addImplicitArgs$1(Typer.scala:2045)
        at dotty.tools.dotc.typer.Typer.adaptNoArgsImplicitMethod$1(Typer.scala:2047)
        at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:2163)
        at dotty.tools.dotc.typer.Typer.adaptInterpolated(Typer.scala:2302)
        at dotty.tools.dotc.typer.Typer.adaptInterpolated(Typer.scala:2291)
        at dotty.tools.dotc.typer.Typer.$anonfun$adapt$3(Typer.scala:1866)
        at dotty.tools.dotc.reporting.trace$.apply(trace.scala:23)
        at dotty.tools.dotc.typer.Typer.$anonfun$adapt$1(Typer.scala:1862)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:1862)
        at dotty.tools.dotc.typer.Implicits$ImplicitSearch.$anonfun$searchImplicits$3(Implicits.scala:789)
        at dotty.tools.dotc.reporting.trace$.apply(trace.scala:23)
        at dotty.tools.dotc.typer.Implicits$ImplicitSearch.$anonfun$searchImplicits$1(Implicits.scala:781)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Implicits$ImplicitSearch.typedImplicit$1(Implicits.scala:781)
        at dotty.tools.dotc.typer.Implicits$ImplicitSearch.rankImplicits$1(Implicits.scala:824)
        at dotty.tools.dotc.typer.Implicits$ImplicitSearch.searchImplicits(Implicits.scala:895)
        at dotty.tools.dotc.typer.Implicits$ImplicitSearch.bestImplicit(Implicits.scala:903)
        at dotty.tools.dotc.typer.Implicits.$anonfun$inferImplicit$4(Implicits.scala:716)
        at dotty.tools.dotc.reporting.trace$.apply(trace.scala:23)
        at dotty.tools.dotc.typer.Implicits.$anonfun$inferImplicit$1(Implicits.scala:711)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Implicits.inferImplicit(Implicits.scala:707)
        at dotty.tools.dotc.typer.Implicits.inferImplicit$(Implicits.scala:707)
        at dotty.tools.dotc.typer.Typer.inferImplicit(Typer.scala:69)
        at dotty.tools.dotc.typer.Implicits.inferImplicitArg(Implicits.scala:625)
        at dotty.tools.dotc.typer.Implicits.inferImplicitArg$(Implicits.scala:553)
        at dotty.tools.dotc.typer.Typer.inferImplicitArg(Typer.scala:69)
        at dotty.tools.dotc.typer.Typer.$anonfun$adaptInterpolated$19(Typer.scala:2021)
        at scala.runtime.Tuple2Zipped$.$anonfun$map$1(Tuple2Zipped.scala:47)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at scala.runtime.Tuple2Zipped$.map$extension(Tuple2Zipped.scala:45)
        at dotty.tools.dotc.typer.Typer.addImplicitArgs$1(Typer.scala:2017)
        at dotty.tools.dotc.typer.Typer.adaptNoArgsImplicitMethod$1(Typer.scala:2047)
        at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:2163)
        at dotty.tools.dotc.typer.Typer.adaptInterpolated(Typer.scala:2302)
        at dotty.tools.dotc.typer.Typer.adaptInterpolated(Typer.scala:2291)
        at dotty.tools.dotc.typer.Typer.$anonfun$adapt$3(Typer.scala:1866)
        at dotty.tools.dotc.reporting.trace$.apply(trace.scala:23)
        at dotty.tools.dotc.typer.Typer.$anonfun$adapt$1(Typer.scala:1862)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:1862)
        at dotty.tools.dotc.typer.Implicits$ImplicitSearch.$anonfun$searchImplicits$3(Implicits.scala:789)
        at dotty.tools.dotc.reporting.trace$.apply(trace.scala:23)
        at dotty.tools.dotc.typer.Implicits$ImplicitSearch.$anonfun$searchImplicits$1(Implicits.scala:781)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Implicits$ImplicitSearch.typedImplicit$1(Implicits.scala:781)
        at dotty.tools.dotc.typer.Implicits$ImplicitSearch.rankImplicits$1(Implicits.scala:824)
        at dotty.tools.dotc.typer.Implicits$ImplicitSearch.searchImplicits(Implicits.scala:895)
        at dotty.tools.dotc.typer.Implicits$ImplicitSearch.bestImplicit(Implicits.scala:903)
        at dotty.tools.dotc.typer.Implicits.$anonfun$inferImplicit$4(Implicits.scala:716)
        at dotty.tools.dotc.reporting.trace$.apply(trace.scala:23)
        at dotty.tools.dotc.typer.Implicits.$anonfun$inferImplicit$1(Implicits.scala:711)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Implicits.inferImplicit(Implicits.scala:707)
        at dotty.tools.dotc.typer.Implicits.inferImplicit$(Implicits.scala:707)
        at dotty.tools.dotc.typer.Typer.inferImplicit(Typer.scala:69)
        at dotty.tools.dotc.typer.Implicits.inferImplicitArg(Implicits.scala:625)
        at dotty.tools.dotc.typer.Implicits.inferImplicitArg$(Implicits.scala:553)
        at dotty.tools.dotc.typer.Typer.inferImplicitArg(Typer.scala:69)
        at dotty.tools.dotc.typer.Typer.$anonfun$adaptInterpolated$19(Typer.scala:2021)
        at scala.runtime.Tuple2Zipped$.$anonfun$map$1(Tuple2Zipped.scala:47)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at scala.runtime.Tuple2Zipped$.map$extension(Tuple2Zipped.scala:45)
        at dotty.tools.dotc.typer.Typer.addImplicitArgs$1(Typer.scala:2017)
        at dotty.tools.dotc.typer.Typer.adaptNoArgsImplicitMethod$1(Typer.scala:2047)
        at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:2163)
        at dotty.tools.dotc.typer.Typer.adaptInterpolated(Typer.scala:2302)
        at dotty.tools.dotc.typer.Typer.$anonfun$adapt$3(Typer.scala:1866)
        at dotty.tools.dotc.reporting.trace$.apply(trace.scala:23)
        at dotty.tools.dotc.typer.Typer.$anonfun$adapt$1(Typer.scala:1862)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:1862)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1700)
        at dotty.tools.dotc.reporting.trace$.apply(trace.scala:23)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1696)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1733)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1745)
        at dotty.tools.dotc.typer.Typer.$anonfun$typedClassDef$1(Typer.scala:1409)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1339)
        at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1627)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1682)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1700)
        at dotty.tools.dotc.reporting.trace$.apply(trace.scala:23)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1696)
        at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1722)
        at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1745)
        at dotty.tools.dotc.typer.Typer.$anonfun$typedPackageDef$1(Typer.scala:1522)
        at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
        at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1509)
        at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1666)
        at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1683)
        at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1700)
        at dotty.tools.dotc.reporting.trace$.apply(trace.scala:23)
        at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1696)
        at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1757)
        at dotty.tools.dotc.typer.FrontEnd.$anonfun$typeCheck$1(FrontEnd.scala:64)
        at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:32)
        at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:62)
        at dotty.tools.dotc.typer.FrontEnd.$anonfun$runOn$7(FrontEnd.scala:93)
        at dotty.tools.dotc.typer.FrontEnd.$anonfun$runOn$7$adapted(FrontEnd.scala:93)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:93)
        at dotty.tools.dotc.Run.$anonfun$compileUnits$3(Run.scala:127)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at dotty.tools.dotc.util.Stats$.trackTime(Stats.scala:47)
        at dotty.tools.dotc.Run.$anonfun$compileUnits$2(Run.scala:125)
        at dotty.tools.dotc.Run.$anonfun$compileUnits$2$adapted(Run.scala:123)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
        at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
        at dotty.tools.dotc.Run.runPhases$1(Run.scala:123)
        at dotty.tools.dotc.Run.$anonfun$compileUnits$1(Run.scala:148)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:86)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:109)
        at dotty.tools.dotc.Run.compileSources(Run.scala:96)
        at dotty.tools.dotc.Run.compile(Run.scala:80)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:25)
        at dotty.tools.dotc.Driver.process(Driver.scala:123)
        at dotty.tools.dotc.Driver.process(Driver.scala:92)
        at dotty.tools.dotc.Driver.process(Driver.scala:104)
        at dotty.tools.dotc.Driver.main(Driver.scala:131)
        at dotty.tools.dotc.Main.main(Main.scala)
</pre>
</details>

--- a/tests/neg/i3542.scala
+++ b/tests/neg/i3542.scala
@@ -1,0 +1,8 @@
+object Test {
+  trait TC[A]
+
+  implicit def case1[F[_]](implicit t: => TC[F[Any]]): TC[String] = ???
+  implicit def case2[G[_]](implicit r: TC[G[Any]]): TC[Int] = ???
+
+  implicitly[TC[Int]] // error: no implicit argument of type TC[Int] found
+}

--- a/tests/pos/i3542-1.scala
+++ b/tests/pos/i3542-1.scala
@@ -1,0 +1,7 @@
+class Foo[T]
+
+object Test {
+  implicit def foo[T](implicit rec: => Foo[T]): Foo[T] = ???
+
+  val bla: Foo[Int] = implicitly[Foo[Int]]
+}

--- a/tests/pos/i3542-2.scala
+++ b/tests/pos/i3542-2.scala
@@ -1,0 +1,20 @@
+trait ::[H, T]
+
+trait Foo[A, R]
+
+trait FooLowPrio {
+  implicit def caseOther[A]: Foo[A, A :: Any] = null
+}
+object Foo extends FooLowPrio {
+  implicit def caseCons[H, HR, T, TR]
+    (implicit // It's a bit artificial: the by name is not required in this example...
+      t: => Foo[T, TR],
+      h: => Foo[H, HR]
+    ): Foo[H :: T, TR] = null
+
+  implicit def caseAny: Foo[Any, Any] = null
+}
+
+object Test {
+  val implicitFoo = implicitly[Foo[Long :: Any, Any]]
+}

